### PR TITLE
ci: lint with stylua

### PR DESCRIPTION
**Pull request by mjlbach**
This PR:

* adds a CI job using the stylua github action that runs stylua --check on PRs
* adds a lint target to the make file for stylua
* formats the nvim lua codebase with stylua
* skips tests for now

If accepted, I will add a commit before merge that adds the format commit to  `.git-blame-ignore-revs` file in the root of neovim 

Which can be consumed with 
`git blame --ignore-revs-file .git-blame-ignore-revs /path/to/whatever.lua`

### CODE REVIEW

**Comment by clason** [comment]
The `utf-8` failures (well, the characters) are intentional since these tests exercise multibyte functionality. This looks like an issue with one of the crates `stylua` uses, so not sure if that can be turned off. But it looks like you can explicitly ignore the offending lines: https://github.com/JohnnyMorganz/StyLua#ignoring-parts-of-a-file

**Comment by mjlbach** [comment]
@clason I don't think that will work, the parsing of the file itself fails due to the invalid utf-8. We can just add these files to an ignorelist https://github.com/JohnnyMorganz/StyLua#filtering-using-styluaignore

**Comment by shadmansaleh** [comment]
```diff
- for char in line:gmatch("[^\128-\191][\128-\191]*") do
+ for char in line:gmatch "[^\128-\191][\128-\191]*" do
```

Seems better with parenthesis better to me . Personal opinion though :)

**Comment by mjlbach** [comment]
@shadmansaleh bring it up with @tjdevries 😆 We talked it over last night and I originally had it the other way. TJ asked upstream for an option to only drop parens when calling with a table, but they said no

**Comment by shadmansaleh** [comment]
> TJ asked upstream for an option to only drop parens when calling with a table, but they said no

Sad ;/ . That would have been better .

**Comment by mjlbach** [comment]
Well, we could always ask again :)

File: runtime/lua/man.lua
Comment (bfredl) [inline runtime/lua/man.lua:1]: if we are going to enforce one style of single line strings, why not ' instead of " ? I find the former easier on the eye.

File: test/functional/lua/buffer_updates_spec.lua
Comment (bfredl) [inline test/functional/lua/buffer_updates_spec.lua:418]: Can this conversion be disabled? I like the semantic distinction between `,` (between fields of a tuple) and `;` (after item in a ordered list of tuples) in places like this.

**Comment by bfredl** [comment]
Would it be insane to fork/vendor stylua into our codebase (if they don't like our idea)? It seems to work pretty well for clint.

File: test/functional/lua/buffer_updates_spec.lua
Comment (mjlbach) [reply test/functional/lua/buffer_updates_spec.lua:418]: No, stylua is an "opinionated" formatter like python's black, so this isn't an option. We could fork or vendor stylua if we wanted to do this

**Comment by shadmansaleh** [comment]
> Would it be insane to fork/vendor stylua into our codebase (if they don't like our idea)? 

Will the rust toolchain be required for neovim then ?

File: runtime/lua/man.lua
Comment (mjlbach) [reply runtime/lua/man.lua:1]: I don't have a strong preference, @tjdevries you?

**Comment by mjlbach** [comment]
@shadmansaleh we can just fork it and roll our own GHA pulling/building the forked stylua without it actually being shipped with neovim.

**Comment by bfredl** [comment]
> Will rust toolchain be required for neovim then ?

I naively expected it to cerntainly be written in lua. :D But no not any more or less than using stylua at all, unforked rust code is still rust code. Also it won't be required unless you actually run the linter locally, just as python3 is not required for neovim just because the clint is written in python3.

**Comment by shadmansaleh** [comment]
> @shadmansaleh we can just fork it and roll our own GHA pulling/building the forked stylua without it actually being shipped with neovim

Awesome.

Is plan to run stylua for formatting or for linting ?
Also I was wondering what current lualint does .

**Comment by mjlbach** [comment]
@shadmansaleh so as it stands the entire codebase is formatted with stylua, and the GHA runs a style check to enforce the formatting, it doesn't reformat after merge or anything.

File: runtime/lua/man.lua
Comment (clason) [reply runtime/lua/man.lua:1]: FWIW, I prefer single quotes as well (don't know why; probably just Stockholm Syndrome by now...)

**Comment by mfussenegger** [comment]
> Would it be insane to fork/vendor stylua into our codebase (if they don't like our idea)? It seems to work pretty well for clint.

I'd be in favour of sticking with the original stylua. Not because I prefer the style,  but because I think *if* we go this route of using an opinionated formatter we shouldn't diverge with our own set of opinions.  For people who work on different projects it would be annoying to have different sets of tooling just to get the styling right

**Comment by clason** [comment]
Especially since the ecosystem (nvim-treesitter, packer) is also starting to use stylua and will be looking at neovim's style settings for guidance.

**Comment by bfredl** [comment]
> , but because I think if we go this route of using an opinionated formatter we shouldn't diverge with our own set of opinions

Not sure I follow, favoring the stylistic choices of stylua over _our own set of opinions_ would be the consequence of doing this.

> For people who work on different projects it would be annoying to have different sets of tooling just to get the styling right

I agree but I disagree with the implication (that seems to be made)  that the defaults of **stylua** should be the default solution. Especially as it, disables some useful idioms, I would hesitate to choose it (at this point) just with the motivation "it is as good as any other style, so let's pick any".

**Comment by clason** [comment]
I think the argument is "if neovim uses stylua _and_ core(-adjacent) plugins use stylua _and_ we are in favor of other plugins looking towards the first two for inspiration/alignment, _then_ we will make their lives significantly easier if they can do so with using our settings in stock `stylua` and don't have to worry about setting up a fork".

Of course, we can also maintain and provide the fork with "neovim defaults" as a wider service, but that would add maintenance burden.

**Comment by bfredl** [comment]
Sure but that's a lot of ands :). We are still in the phase to decide if we want stylua _at all_ (even though most of us seem positive to the idea of _something_ that's stricter to lualint). Seeing this a choice that implicates stuff for the entire ecosystem would make me __more__ concerned of the concrete style choices, and less willing to just accept decisions of a rust crate I haven't heard about since 12 hours ago as the natural default. (again, in the short term. they seem not to _completely_ ruled out the possibility to make changes).

> Of course, we can also maintain and provide the fork with "neovim defaults" as a wider service, but that would add maintenance burden.

I mentioned forking simply on the basis than we already do this for `make clint` which hasn't seemed too much of a maintenance burden. (It is not the prettiest py3 code I have read, but the 3-4 times we have needed to fix/change something we have done so with  reasonable effort). Though this was before the scope of the discussion was increased the entire ecosystem. I would (very naively) assume a lua formatter to be available written in lua+lpeg (not sure if that changes the perceived maintenance burden), and I would look out for other options before making this decision, if it is to be made with this scope.

File: runtime/lua/man.lua
Comment (shadmansaleh) [reply runtime/lua/man.lua:1]: I don't have a strong opinion here . But I'm also inclined towards single quotes.

**Comment by clason** [comment]
All reasonable points, and "the ecosystem (even nvim-lspconfig) is on its own" is a valid option here.

FWIW, `luarocks` has a comprehensive [style guide](https://github.com/luarocks/lua-style-guide) and recommends [`luacheck`](https://github.com/mpeterv/luacheck) as a linter/style checker. This doesn't format, though.

**Comment by justinmk** [comment]
Really great discussion/analysis here. 

+1 for single quotes, that also matches python and it's easier to type on US layout.

If https://github.com/Koihik/LuaFormatter looks maintained, works well, and is closer to our build chain (so it's easier for contributors to use locally), that gives it an edge over stylua.

This is a reversible decision though, easy to change to another too later. So if we have something working now let's use that. As long as it does single quotes :D

File: .github/workflows/ci.yml
Comment (justinmk) [inline .github/workflows/ci.yml:89]: nit: why mention the tool (implementation detail) in the step name ?

Also isn't this doing formatting, not linting?
```suggestion
      - name: Enforce code style
```

File: .stylua.toml
Comment (justinmk) [inline .stylua.toml:1]: maybe we should start keeping these files in [config/](https://github.com/neovim/neovim/tree/master/config) ?

**Review by justinmk — APPROVED** [review]

**Comment by tjdevries** [comment]
I will try and comment this week. I do not like default options for stylua :) I am willing to do single quotes. I like dropping parenths for the functions. column width 120 is our current lua lint width and I think it's good (it's not the same as our C guidelines, if I recall).

Why does what tech is used matter to using it for the project? can't we just download a release or something to run it?

(Only briefly read things, will look more later)

**Comment by bfredl** [comment]
> Why does what tech is used matter to using it for the project? can't we just download a release or something to run it?

_Ideally_ yes, but it seems stylua hardcodes a lot more choices (c.f. LuaFormatter), and is not keen to add more configuration, so it seems we would need to maintain a fork just for basic option changes.. While as LuaFormatter more flexible so we can get closer to our preferences already out of the box (and presumably would be more willing to accept our patches, if needed, so we don't need to carry a fork in this case).

File: .github/workflows/ci.yml
Comment (mjlbach) [reply .github/workflows/ci.yml:89]: Happy to change the name, I put the name in the tool so it's more obvious which lint users should run locally, but we could add a `make format` command to automatically run our lua formatter of choice.

Re: lint vs. format, it's currently only checking the formatting in CI, it doesn't actually format the code (unless invoked locally by users without the --check flag)

File: .stylua.toml
Comment (mjlbach) [reply .stylua.toml:1]: Yes, I didn't like that it clobbered the root either. @tjdevries and I have been informally using 120, but we can also start being stricter on the lua code (or less strict on the C code) pending this discussion :)

**Comment by mjlbach** [comment]
I also evaluated https://github.com/Koihik/LuaFormatter, I think the issues is the raised concerns (formatting of tables, dropping of parens around only functions taking a single table as an argument) are not currently options in LuaFormatter (either, at least I couldn't see them when I had looked). Given the available options already, the author might be more amenable to adding our requests than stylua is to avoid the maintenance burden of us forking.

**Comment by mjlbach** [comment]
So I updated the (non-rebased, since I don't think we'll merge this soon) PR with single quote/out of tree stylua.toml, one annoying thing is stylua does not let you put the styluaignore out of tree, so maybe we can bug upstream about that, since there's a flag for the config file

`stylua --config-path ci/.stylua.toml  .   `

**Comment by lithammer** [comment]
I have to admit I'm against putting `.stylua.toml` in a subdirectory. It breaks automatic config discovery which in turn breaks editor/tooling integration.

Configuring flags on a per-project basis is a massive hassle. I'd rather just have it work out-of-the-box (i.e put it in the root of the project).

However I understand following existing conventions. But as it is now I would have to disable automatic formatting and run `make format` after I'm done editing. Or temporarily have a super messy diff because of formatting with the wrong settings and run `make format` before committing.

**Comment by mjlbach** [comment]
Yes, I'm torn too. It is kinda annoying each time I accidentally run stylua without passing the config path on this branch haha.

**Comment by justinmk** [comment]
> Yes, I'm torn too. It is kinda annoying each time I accidentally run stylua without passing the config path on this branch haha.

let's forget about putting it in config/ then.

**Comment by mjlbach** [comment]
We could just add aliases for the command in make lint/format though to work around it

**Comment by justinmk** [comment]
> We could just add aliases for the command in make lint/format though to work around it

yeah, though https://github.com/neovim/neovim/pull/14987#issuecomment-880803700 is an argument I hadn't considered.

In fact I did something similar in 67305ffb5de865f872898c03cbacd189c9e9aa53

**Comment by mjlbach** [comment]
I brought this PR up to date, but I don't think status-quo has changed on how much we agree with the available formatting options in stylua. I would like to move forward with *something* so I can try doing what dundard did and going through https://github.com/Koihik/LuaFormatter to see if we can tweek the options until everyone is happy. The other option is just disallowing stylua in the offending files.

File: .stylua.toml
Comment (Shatur) [inline .stylua.toml:6]: `column_width = 120`, `line_endings = "Unix"` and `no_call_parentheses = false` are [defaults](https://github.com/JohnnyMorganz/StyLua#finding-the-configuration) in stylua. You may omit it as suggested in the readme.

Comment (gpanders) [reply .stylua.toml:6]: -1, defaults can change and it's nice to be explicit

**Comment by MunifTanjim** [comment]
Just putting a note here about the `no_call_parentheses`.

Having this turned on has some side-effects.

- For someone who's reading the code, sometimes functions are called without parentheses (with single string/table param) and the other times they are called with parentheses (with multiple params). Also if a function has only single required param and some optional params, that might be called in various places sometimes with/without parentheses. That feels unpleasant to me.

- For someone who's writing the code, say a function used to take a single param and now you're adding another param, you need to update all the callsites where the second param needs to be passed... but you not only need to add the second param, you need to wrap it in parentheses too.

Granted this are tiny issues and very easy to overlook... but why bother? 🤷🏼‍♂️ Keeping it uniform in all the places comes with zero cost. Unless it's a shell language, functions are called with parentheses almost in every languages. Even somebody uninitiated with lua syntax sugar can instantly recognize a function call by seeing those parentheses.

**Comment by mjlbach** [comment]
Please update your comments on the diff, specifically those who had objects @bfredl @justinmk 

Stylua added the long awaited call table option: https://github.com/JohnnyMorganz/StyLua/commit/6272fdfd1bf3cae4b183b73f71191ddb8a3beb5b

It's not cut in a release yet. I'll try to add the lint CI job.

**Comment by mjlbach** [comment]
This is semi-blocked by wanting to wait for the next stylua release (even if this is the way to go).

File: test/functional/api/version_spec.lua
Comment (bfredl) [inline test/functional/api/version_spec.lua]: This is a bit unfortunate (and follows no comprehensible rule). Could it be changed to not do that?

Comment (clason) [reply test/functional/api/version_spec.lua]: No.

Comment (clason) [reply test/functional/api/version_spec.lua]: (except by increasing the admissible line length)

Comment (clason) [reply test/functional/api/version_spec.lua]: I'd have to check, but I assume it follows the same rule as for argument lists (`foo(a,b,c,d)`): if the full list in parentheses does not fit into the same line, every item gets its own line.

Comment (clason) [reply test/functional/api/version_spec.lua]: https://github.com/JohnnyMorganz/StyLua/issues/139

File: src/nvim/auevents.lua
Comment (bfredl) [inline src/nvim/auevents.lua:3]: perhaps exclude this file in its entirety?

File: src/nvim/generators/gen_api_dispatch.lua
Comment (bfredl) [inline src/nvim/generators/gen_api_dispatch.lua:124]: an example in non-text code of the same issue. the only thing wrong here is a hanging `..` which should be on the next line. Ideally it should be able to fix this formatting issue without completely _deformatting_ the entire statement at the same time.

**Comment by CppCXY** [comment]
I think [EmmyluaCodeStyle](https://github.com/CppCXY/EmmyLuaCodeStyle) is an option to consider, and it has been integrated into sumneko_lua ls

**Comment by mjlbach** [comment]
Given the activity on https://github.com/neovim/neovim/pull/17547, I'm happy to revive this PR this weekend.

**Comment by clason** [comment]
Bump -- I'd really like to start using a formatter for Lua files. If I may summarize the discussion (including my own opinion), I believe we have broad (if not total) consensus on the following plan:

1. we _want_ a Lua linter;
2. we prefer Stylua, as it's simple to configure, comes with a robust GH Action, and is already widely used throughout the ecosystem;
3. there's somewhat of a (vocal) majority for unconditionally _keeping_ call parens, as this reduces friction for new contributors;
4. we want to be conservative in which files to lint; I believe the following would be uncontroversial:
 a) `runtime/**/*.lua`
 b) `src/nvim/lua/vim.lua`

Points 3 and 4 can be revisited later.

@mjlbach as a "farewell gift", could you rebase the PR and make these changes?

**Comment by CppCXY** [comment]
It seems that you have decided to use `stylua`, maybe I can close my pr. by the way, my tool is also a code-style linter.

**Comment by CppCXY** [comment]
@clason 
My tool is also powerful lint tool. it might also be worth considering

**Comment by clason** [comment]
I don't doubt it, and it looks even more powerful (configurable). But from what I've seen so far, Stylua is _easier_ to integrate _for us_ (due to the reasons I listed above).

**Comment by CppCXY** [comment]
ok, thanks for the discussion

**Pull request closed without merging**
